mr/src/main/java/org/apache/iceberg/mr/hive/HiveSchemaUtil.java

```diff
@@ -0,0 +1,171 @@
+      List<Types.NestedField> result = new ArrayList<>(names.size());
+      for (int i = 0; i < names.size(); ++i) {
+        result.add(visit(names.get(i), typeInfos.get(i)));
+      }
+
+      return result;
+    }
+
+    private Types.NestedField visit(String name, TypeInfo typeInfo) {
+      switch (typeInfo.getCategory()) {
+        case PRIMITIVE:
+          switch (((PrimitiveTypeInfo) typeInfo).getPrimitiveCategory()) {
+            case FLOAT:
+              return Types.NestedField.optional(id++, name, Types.FloatType.get());
+            case DOUBLE:
+              return Types.NestedField.optional(id++, name, Types.DoubleType.get());
+            case BOOLEAN:
+              return Types.NestedField.optional(id++, name, Types.BooleanType.get());
+            case BYTE:
+            case SHORT:
+            case INT:
+              return Types.NestedField.optional(id++, name, Types.IntegerType.get());
+            case LONG:
+              return Types.NestedField.optional(id++, name, Types.LongType.get());
+            case BINARY:
+              return Types.NestedField.optional(id++, name, Types.BinaryType.get());
+            case STRING:
+            case VARCHAR:
+              return Types.NestedField.optional(id++, name, Types.StringType.get());
+            case CHAR:
+              Types.FixedType fixedType = Types.FixedType.ofLength(((CharTypeInfo) typeInfo).getLength());
+              return Types.NestedField.optional(id++, name, fixedType);
+            case TIMESTAMP:
+              return Types.NestedField.optional(id++, name, Types.TimestampType.withZone());
+            case DATE:
+              return Types.NestedField.optional(id++, name, Types.DateType.get());
+            case DECIMAL:
+              DecimalTypeInfo decimalTypeInfo = (DecimalTypeInfo) typeInfo;
+              Types.DecimalType decimalType =
+                  Types.DecimalType.of(decimalTypeInfo.precision(), decimalTypeInfo.scale());
+              return Types.NestedField.optional(id++, name, decimalType);
+            // TODO: In Hive3 we have TIMESTAMPLOCALTZ
+            default:
+              throw new IllegalArgumentException("Unknown primitive type " +
+                  ((PrimitiveTypeInfo) typeInfo).getPrimitiveCategory());
+          }
+        case STRUCT:
+          StructTypeInfo structTypeInfo = (StructTypeInfo) typeInfo;
+          List<Types.NestedField> fields =
+              visit(structTypeInfo.getAllStructFieldNames(), structTypeInfo.getAllStructFieldTypeInfos());
+          Types.StructType structType = Types.StructType.of(fields);
+          return Types.NestedField.optional(id++, name, structType);
+        case MAP:
+          MapTypeInfo mapTypeInfo = (MapTypeInfo) typeInfo;
+          Types.NestedField keyField = visit(name + "_key", mapTypeInfo.getMapKeyTypeInfo());
+          Types.NestedField valueField = visit(name + "_value", mapTypeInfo.getMapValueTypeInfo());
```

Comment:
Would it make sense to have a method for `TypeInfo` so you don't have to create an artificial name and then discard the field information that isn't data type?

Reply:
Done